elrond-wasm/src/storage/mappers/linked_list_mapper.rs

```diff
@@ -0,0 +1,240 @@
+
+const NULL_ENTRY : u32 = 0;
+const LEN_IDENTIFIER : &[u8] = b".len";
+const NEW_KEY_IDENTIFIER : &[u8] = b".new";
+const FRONT_IDENTIFIER : &[u8] = b".front";
+const BACK_IDENTIFIER : &[u8] = b".back";
+const NODE_IDENTIFIER : &[u8] = b"*";
+const VALUE_IDENTIFIER : &[u8] = b"&";
+
+/// A unit of balance, usually stake.
+/// Contains a description of the source/intent of the funds, together with a balance.
+#[derive(TopEncode, TopDecode, PartialEq, Clone, Copy)]
+pub struct Node {
+    pub previous: u32,
+    pub next: u32
+}
+
+/// Manages a doubly-linked list of items
+pub struct LinkedListMapper<SA, T>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+	T: TopEncode + TopDecode + 'static,
+{
+	api: SA,
+	main_key: BoxedBytes,
+	_phantom: core::marker::PhantomData<T>,
+}
+
+impl<SA, T> StorageMapper<SA> for LinkedListMapper<SA, T>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+	T: TopEncode + TopDecode,
+{
+	fn new(api: SA, main_key: BoxedBytes) -> Self {
+		let result = LinkedListMapper {
+			api,
+			main_key,
+			_phantom: PhantomData,
+		};
+		result.set_len(0);
+		result.set_front(NULL_ENTRY);
+		result.set_back(NULL_ENTRY);
+		result
+	}
+}
+
+impl<SA, T> LinkedListMapper<SA, T>
+where
+	SA: StorageReadApi + StorageWriteApi + ErrorApi + Clone + 'static,
+	T: TopEncode + TopDecode,
+{
+	fn build_index_named_key(&self, name : &[u8], index : u32) -> BoxedBytes {
+		BoxedBytes::from_concat(&[self.main_key.as_slice(), name, &index.to_be_bytes()])
+	}
+
+	fn build_name_key(&self, name : &[u8]) -> BoxedBytes {
+		BoxedBytes::from_concat(&[self.main_key.as_slice(), name])
+	}
+
+	fn get_u32(&self, name : &[u8]) -> u32 {
+		storage_get(self.api.clone(), self.build_name_key(name).as_slice())
+	}
+
+	fn set_u32(&self, name : &[u8], value : u32) {
+		storage_set(self.api.clone(), self.build_name_key(name).as_slice(), &value);
+	}
+
+	fn get_len(&self) -> u32 {
+		self.get_u32(LEN_IDENTIFIER)
+	}
+
+	fn set_len(&self, len : u32) {
+		self.set_u32(LEN_IDENTIFIER, len);
+	}
+
+	fn generate_new_key(&self) -> u32 {
+		let new_key = self.get_u32(NEW_KEY_IDENTIFIER) + 1;
+		self.set_u32(NEW_KEY_IDENTIFIER, new_key);
+		new_key
+	}
+
+	fn get_front(&self) -> u32 {
+		self.get_u32(FRONT_IDENTIFIER)
+	}
+
+	fn set_front(&self, index : u32) {
+		self.set_u32(FRONT_IDENTIFIER, index);
+	}
+
+	fn get_back(&self) -> u32 {
+		self.get_u32(BACK_IDENTIFIER)
+	}
+
+	fn set_back(&self, index : u32) {
+		self.set_u32(BACK_IDENTIFIER, index);
+	}
+
+	fn get_node(&self, index : u32) -> Node {
+		storage_get(self.api.clone(), self.build_index_named_key(NODE_IDENTIFIER, index).as_slice())
+	}
+
+	fn set_node(&self, index : u32, item : Node) {
+		storage_set(self.api.clone(), self.build_index_named_key(NODE_IDENTIFIER, index).as_slice(), &item);
+	}
+
+	fn clear_node(&self, index : u32) {
+		storage_set(self.api.clone(), self.build_index_named_key(NODE_IDENTIFIER, index).as_slice(), &BoxedBytes::empty());
```

Comment:
storage_set(self.api.clone(), self.build_index_named_key(NODE_IDENTIFIER, index).as_slice(), &[u8; 0][...]);

Reply:
i think

Reply:
I will actually some better base api for mappers, so you can also leave like this for now

Reply:
Left as-is